File: chapter3/python/3.2 StackMin.py
# Solution


class MinStack:
    def __init__(self):
        self.items = []
        self.min_arr = []
        self.min_num = 0x3f3f3f3f

    def size(self):
        return len(self.items)

    def push(self, item):
        self.items.append(item)
        self.min_num = min(self.min_num, item)
        self.min_arr.append(self.min_num)

    def pop(self):
        if self.size() == 0:
            return None
        else:
            self.min_arr.pop()
            return self.items.pop()

    def min(self):
        return self.min_arr[-1]


# TESTS
MyStack = MinStack()

MyStack.push(40)
print(MyStack.min())  # 40
MyStack.push(20)
print(MyStack.min())  # 20
MyStack.push(30)
print(MyStack.min())  # 20
MyStack.push(10)
print(MyStack.min())  # 10
MyStack.pop()
print(MyStack.min())  # 20
MyStack.pop()
print(MyStack.min())  # 20
MyStack.pop()
print(MyStack.min())  # 40

# This solution is O(1) time and O(n) space

File: chapter3/javascript/3.2 StackMin.js
class StackMin {
    constructor() {
        this.values = [];
        this.min_arr = [];
        this.min_num = 0x3f3f3f3f;
    }

    size() {
        return this.values.length;
    }

    isEmpty() {
        return this.values.length === 0;
    }

    push(val) {
        this.values.push(val);
        this.min_num = Math.min(this.min_num, val);
        this.min_arr.push(this.min_num)
    }

    pop() {
        this.min_arr.pop();
        return this.values.pop();
    }

    top() {
        return this.values[this.values.length - 1];
    }

    toArray() {
        return this.values;
    }

    min() {
        return this.min_arr[this.min_arr.length - 1];
    }
}

let my_stack = new StackMin();

console.log(my_stack.isEmpty()); // true
console.log(my_stack.pop()); // undefined
console.log(my_stack.size()); // 0
my_stack.push(40);
console.log(my_stack.min()); // 40
my_stack.push(20);
console.log(my_stack.min()); // 20
my_stack.push(30);
console.log(my_stack.min()); // 20
my_stack.push(10);
console.log(my_stack.min()); // 10
my_stack.pop();
console.log(my_stack.min()); // 20
my_stack.pop();
console.log(my_stack.min()); // 20
my_stack.pop();
console.log(my_stack.min()); // 40

// This solution is O(1) time and O(n) space